internal/extproc/translator/openai_awsbedrock.go

```diff
@@ -152,113 +180,194 @@ func (o *openAIToAWSBedrockTranslatorV1ChatCompletion) openAIToolsToBedrockToolC
+}
+
+// openAIMessageToBedrockMessageRoleTool converts openai tool role message
+func (o *openAIToAWSBedrockTranslatorV1ChatCompletion) openAIMessageToBedrockMessageRoleTool(
+	openAiMessage *openai.ChatCompletionMessageParamUnion,
+) (*awsbedrock.Message, error) {
+	message := openAiMessage.Value.(openai.ChatCompletionToolMessageParam)
+	return &awsbedrock.Message{
+		// bedrock does not support tool role, merging to the user role
+		Role: awsbedrock.ConversationRoleUser,
+		Content: []*awsbedrock.ContentBlock{
+			{
+				ToolResult: &awsbedrock.ToolResultBlock{
+					Content: []*awsbedrock.ToolResultContentBlock{
+						{
+							Text: message.Content.Value.(*string),
+						},
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+// openAIMessageToBedrockMessage converts openai ChatCompletion messages to aws bedrock messages
+func (o *openAIToAWSBedrockTranslatorV1ChatCompletion) openAIMessageToBedrockMessage(openAIReq *openai.ChatCompletionRequest,
 	bedrockReq *awsbedrock.ConverseInput,
 ) error {
 	// Convert Messages.
 	bedrockReq.Messages = make([]*awsbedrock.Message, 0, len(openAIReq.Messages))
 	for _, msg := range openAIReq.Messages {
 		switch msg.Type {
 		case openai.ChatMessageRoleUser:
-			message := msg.Value.(openai.ChatCompletionUserMessageParam)
-			if _, ok := message.Content.Value.(string); ok {
-				bedrockReq.Messages = append(bedrockReq.Messages, &awsbedrock.Message{
-					Role: msg.Type,
-					Content: []*awsbedrock.ContentBlock{
-						{Text: ptr.To(message.Content.Value.(string))},
-					},
-				})
-			} else {
-				if contents, ok := message.Content.Value.([]openai.ChatCompletionContentPartUserUnionParam); ok {
-					chatMessage := &awsbedrock.Message{Role: msg.Type}
-					chatMessage.Content = make([]*awsbedrock.ContentBlock, 0, len(contents))
-					for _, contentPart := range contents {
-						if contentPart.TextContent != nil {
-							textContentPart := contentPart.TextContent
-							chatMessage.Content = append(chatMessage.Content, &awsbedrock.ContentBlock{
-								Text: &textContentPart.Text,
-							})
-						} else if contentPart.ImageContent != nil {
-							imageContentPart := contentPart.ImageContent
-							parts := strings.Split(imageContentPart.ImageURL.URL, ",")
-							if len(parts) == 2 {
-								formatPart := strings.Split(parts[0], ";")[0]
-								format := strings.TrimPrefix(formatPart, "data:image/")
-								chatMessage.Content = append(chatMessage.Content, &awsbedrock.ContentBlock{
-									Image: &awsbedrock.ImageBlock{
-										Format: format,
-										Source: awsbedrock.ImageSource{
-											Bytes: []byte(parts[1]),
-										},
-									},
-								})
-							} else {
-								return fmt.Errorf("unexpected image data url")
-							}
-						}
-					}
-					bedrockReq.Messages = append(bedrockReq.Messages, chatMessage)
-				} else {
-					return fmt.Errorf("unexpected content type for user message")
-				}
+			bedrockMessages, err := o.openAIMessageToBedrockMessageRoleUser(&msg)
```

Comment:
this escapes the message... instead make it the pointer at the beginning of the loop in the first place.